lib/improver/standardise.py

```diff
@@ -0,0 +1,442 @@
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Plugin to regrid cube data and standardise metadata"""
+
+import warnings
+import numpy as np
+
+import iris
+from iris.analysis import Nearest, Linear
+from iris.exceptions import CoordinateNotFoundError
+from scipy.interpolate import griddata
+
+from improver import BasePlugin
+from improver.metadata.amend import amend_attributes
+from improver.metadata.check_datatypes import (
+    check_cube_not_float64, check_time_coordinate_metadata)
+from improver.metadata.constants.time_types import (
+    TIME_COORD_NAMES, TIME_REFERENCE_DTYPE, TIME_REFERENCE_UNIT,
+    TIME_DTYPE, TIME_UNIT)
+from improver.threshold import BasicThreshold
+from improver.utilities.cube_checker import spatial_coords_match
+from improver.utilities.spatial import OccurrenceWithinVicinity
+
+
+def grid_contains_cutout(grid, cutout):
+    """
+    Check that a spatial cutout is contained within a given grid
+
+    Args:
+        grid (iris.cube.Cube):
+            A cube defining a data grid
+        cutout (iris.cube.Cube):
+            The cutout to search for within the grid
+
+    Returns:
+        bool:
+            True if cutout is contained within grid, False otherwise
+    """
+    if spatial_coords_match(grid, cutout):
+        return True
+
+    # check whether "cutout" coordinate points match a subset of "grid"
+    # points on both axes
+    for axis in ['x', 'y']:
+        grid_coord = grid.coord(axis=axis)
+        cutout_coord = cutout.coord(axis=axis)
+        # check coordinate metadata
+        if (cutout_coord.name() != grid_coord.name() or
+                cutout_coord.units != grid_coord.units or
+                cutout_coord.coord_system != grid_coord.coord_system):
+            return False
+
+        # search for cutout coordinate points in larger grid
+        cutout_start = cutout_coord.points[0]
+        if not np.any([np.isclose(cutout_start, grid_point)
+                       for grid_point in grid_coord.points]):
+            return False
+
+        start = list(grid_coord.points).index(cutout_start)
```

Comment:
Also: I should catch the IndexError here that might arise if my cutout domain goes over the edge of my grid (and return False rather than raising the error).

Reply:
Eurgh actually it raises a ValueError.  This is non-obvious numpy slicing behaviour so I've included a comment explaining what's happening.